chainer_pytorch_migration/parameter.py

```diff
@@ -16,12 +30,94 @@ class LinkAsTorchModel(nn.Module):
     '''
 
     def __init__(self, link):
-        super(LinkAsTorchModel, self).__init__()
-        for n, p in sorted(link.namedparams()):
-            self.__setattr__(n, ChainerParameter(p))
+        super().__init__()
+
+        for name, child in _named_children(link):
+            child_module = LinkAsTorchModel(child)
+            setattr(self, name, child_module)
+        for name, param in sorted(_named_params(link)):
+            setattr(self, name, ChainerParameter(param))
+
+        self.link = link
+
+    def forward(self, *input):
+        # The computation graph should be done in Chainer.
+        # Forward converts the input tensors to numpy/cupy arrays
+        # as accepted by Chainer.
+        # The return value should be a tensor as well.
+        input = [cpm.tensor.asarray(x) if isinstance(x, torch.Tensor)
+                 else x for x in input]
+        outputs = self.link.forward(*input)
+        ret = self.__as_tensor(outputs)
+        return ret
+
+    def __as_tensor(self, value):
+        if isinstance(value, tuple):
+            return tuple(self.__as_tensor(x) for x in value)
+        if isinstance(value, list):
+            return [self.__as_tensor(x) for x in value]
+        if isinstance(value, chainer.Variable):
+            return _ChainerTensor(value)
+        return value
+
+
+class Optimizer(torch.optim.Optimizer):
+
+    def __init__(self, base_optimizer):
+        assert isinstance(base_optimizer, torch.optim.Optimizer)
+        super().__init__(base_optimizer.param_groups, base_optimizer.defaults)
+        self._base_optimizer = base_optimizer
+
+    def __getattr__(self, name):
+        if name in ('step', 'zero_grad', '_base_optimizer'):
+            return object.__getattribute__(self, name)
+        return getattr(self._base_optimizer, name)
+
+    def step(self, closure=None):
+        for param_group in self._base_optimizer.param_groups:
+            for param in param_group['params']:
+                assert isinstance(param, ChainerParameter)
+                param.grad.copy_(torch.tensor(param._param.grad))
```

Comment:
could the grad temporal copy be avoided by doing
`param.grad.copy_(cpm.astensor(param._param.grad))`?

Reply:
👍 Thanks, fixed in b968d1767a714d2e206806d67fc87306096f5209